File: frontend/src/types/pnp.ts
export interface PnPComponent {
  designator: string;
  value: string | null;
  package: string | null;
  x: number;
  y: number;
  rotation: number;
  side: string;
}

export interface PnPParseResponse {
  format: string;
  components: PnPComponent[];
  bounds: { min_x: number; min_y: number; max_x: number; max_y: number };
  total_count: number;
  top_count: number;
  bottom_count: number;
}

export interface CalibrationResidual {
  file_x: number;
  file_y: number;
  machine_x: number;
  machine_y: number;
  predicted_x: number;
  predicted_y: number;
  error_mm: number;
}

export interface TransformInfo {
  rotation_deg: number;
  scale_x: number;
  scale_y: number;
  offset_x: number;
  offset_y: number;
  rms_error: number;
  residuals: CalibrationResidual[];
  warnings: string[];
  matrix: { a: number; b: number; c: number; d: number; tx: number; ty: number };
}

export interface CalibrateResponse {
  components: PnPComponent[];
  transform: TransformInfo;
}

File: backend/app/schemas/pnp.py
"""Pydantic schemas for the Pick & Place converter."""

from pydantic import BaseModel


class PnPComponent(BaseModel):
    designator: str
    value: str | None = None
    package: str | None = None
    x: float
    y: float
    rotation: float = 0.0
    side: str = "top"


class PnPParseResponse(BaseModel):
    format: str
    components: list[PnPComponent]
    bounds: dict
    total_count: int
    top_count: int
    bottom_count: int


class CalibrationPoint(BaseModel):
    x: float
    y: float


class CalibrateRequest(BaseModel):
    components: list[PnPComponent]
    file_points: list[CalibrationPoint]
    machine_points: list[CalibrationPoint]


class CalibrateResponse(BaseModel):
    components: list[PnPComponent]
    transform: dict  # {"rotation_deg": float, "scale": float, "offset_x": float, "offset_y": float}


class ExportRequest(BaseModel):
    components: list[PnPComponent]
    side_filter: str = "all"  # "top", "bottom", "all"


class PnPSummary(BaseModel):
    total: int
    top: int
    bottom: int
    unique_packages: list[str]
    unique_values: list[str]
    package_counts: dict[str, int]
